src/classLibrary/interfaces/interfaces.ts

```diff
@@ -0,0 +1,93 @@
+import { of, Union } from "ts-union";
+
+type Syntax = [string,string] // [open, close] , [ ["~", "~"] , ["{{"], ["}}"]]
+
+interface CyanObject {
+	globs: Glob[] | Glob;
+	copyOnly: Glob[] | Glob; // Doesn't go through parser
+	variable: object; // { a: "Roses", b: { c: "Violets", d: "please" } }
+	flags: object; // { remove: { one: true, two: false } }
+	guid: string[] | string; // ["asdf-asdf-asdf-asdf", "1234'1234'1234"]
+	syntax: Syntax[] | Syntax;
+	plugins: { [s: string]: string[] }; // {"api.cyanprint.dev": ["npm, c#", "github"]}
+	comments: string[] | string; // remove these strings if cyanprint templates starts with them
+	pluginData: object | null | undefined;
+}
+
+// Parsed version
+interface CyanSafe {
+	globs: Glob[];
+	copyOnly: Glob[];
+	variable: object;
+	flags: object;
+	guid: string[];
+	syntax: Syntax[];
+	plugins: { [s: string]: string[] };
+	comments: string[];
+	pluginData: object;
+}
+
+interface ICyanParser {
+	Parse(p: Partial<CyanObject>): CyanSafe;
+}
+
+interface IFileSystemInstanceMetadata {
+	sourceAbsolutePath: string;
+	destinationAbsolutePath: string;
+	relativePath: string;
+}
+
+const FileContent = Union({
+	String: of<string>(),
+	Buffer: of<Buffer>(),
+});
+
+type FileContent = typeof FileContent.T;
+
+interface FileSystemInstance {
+	metadata: IFileSystemInstanceMetadata;
+	content: FileContent;
+	parse: boolean;
```

Comment:
What is parse for again? I forgotten. Also, following the introduction of union types, maybe we can make the file system instance a union type of Directory and File?

Reply:
Parse was just a flag to decide whether the file should be parsed or not, if false then it will just be left alone (I think rationale is for maybe Licenses(?) or certain dotfiles that the user can define beforehand)

Reply:
ohh isit for the `copyOnly` flag?

Reply:
so that is doesn't parse but copies only. is this logic checked on the resolver/strategy level alr?

Reply:
Also yeah that makes sense, we can use a union type of Directory and File

Reply:
Yeah its for the `copyOnly` flag. The logic is being checked already, I don't think I included test cases for it yet so I'll add it in

Reply:
@kirinnee Would the method signature of `ParsingStrategy` change? For e.g., currently `Count` takes in `files: FileSystemInstance[]`. 

After adding a union type of Directory and File, assuming called `VirtualFileSystemInstance`, would the method for `Count` change to take in `files: VirtualFileSystemInstance[]`

Reply:
This is with context to the other two methods for `ParsingStrategy` which are `ResolveContents` which theoretically should only resolve the contents of `Files` and not `Directory`

Reply:
> @kirinnee Would the method signature of `ParsingStrategy` change? For e.g., currently `Count` takes in `files: FileSystemInstance[]`. 
> 
> After adding a union type of Directory and File, assuming called `VirtualFileSystemInstance`, would the method for `Count` change to take in `files: VirtualFileSystemInstance[]`

Yes. Count will count all VFS instance.

Reply:
> This is with context to the other two methods for `ParsingStrategy` which are `ResolveContents` which theoretically should only resolve the contents of `Files` and not `Directory` 

Yes again. Resolve content will only read Files. However, you might want to think about how the parent method calls it in a pipeline, and decide whehter it should only take in VFS [] or File[]

Reply:
I mean that's where idk. Strictly speaking if the parser is supposed to `ResolveContents`, then it should just take in File[], but I'm thinking since its other methods are accepting VFS already, might as well accept VFS as well

Reply:
Yup. Taking in VFS violates law of demeter. But I think taking in VFS and returning VFS is easier on the pipeline. Just my 2cents. I'm ok w either implementations.